plugins/scaffolder-backend-module-gitlab/src/actions/createProjectDeployToken.ts

```diff
@@ -0,0 +1,111 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { createTemplateAction } from '@backstage/plugin-scaffolder-backend';
+import { Gitlab } from '@gitbeaker/node';
+import { ScmIntegrationRegistry } from '@backstage/integration';
+import { InputError } from '@backstage/errors';
+import { DeployTokenScope } from '@gitbeaker/core/dist/types/templates/ResourceDeployTokens';
+import { parseRepoHost } from '../util';
+
+export const createGitlabProjectDeployToken = (options: {
+  integrations: ScmIntegrationRegistry;
+}) => {
+  const { integrations } = options;
+  return createTemplateAction<{
+    repoUrl: string;
+    projectId: string | number;
+    name: string;
+    username: string;
+    scopes: string[];
+    token?: string;
+  }>({
+    id: 'gitlab:create-project-deploy-token',
+    schema: {
+      input: {
+        required: ['projectId'],
+        type: 'object',
+        properties: {
+          repoUrl: {
+            title: 'Repository Location',
+            type: 'string',
+          },
+          projectId: {
+            title: 'Project ID',
+            type: 'string | number',
+          },
+          name: {
+            title: 'Deploy Token Name',
+            type: 'string',
+          },
+          username: {
+            title: 'Deploy Token Username',
+            type: 'string',
+          },
+          scopes: {
+            title: 'Scopes',
+            type: 'array',
+          },
+          token: {
+            title: 'Authentication Token',
+            type: 'string',
+            description: 'The token to use for authorization to GitLab',
+          },
+        },
+      },
+    },
+    async handler(ctx) {
+      ctx.logger.info(`Creating Token for Project "${ctx.input.projectId}"`);
+      const { repoUrl, projectId, name, username, scopes } = ctx.input;
+
+      const host = parseRepoHost(repoUrl);
+      const integrationConfig = integrations.gitlab.byHost(host);
+
+      if (!integrationConfig) {
+        throw new InputError(
+          `No matching integration configuration for host ${host}, please check your integrations config`,
+        );
+      }
+
+      const token = ctx.input.token || integrationConfig.config.token!;
+      const tokenType = ctx.input.token ? 'oauthToken' : 'token';
+
+      if (tokenType === 'oauthToken') {
+        throw new InputError(`OAuth Token is currently not supported`);
+      }
+
+      const api = new Gitlab({
+        host: integrationConfig.config.baseUrl,
+        token: token,
+      });
+
+      const deployToken = await api.ProjectDeployTokens.add(
+        projectId,
+        name,
+        scopes as DeployTokenScope[],
+        {
+          username: username,
+        },
+      );
+
+      const string = JSON.stringify(deployToken);
+      const result = JSON.parse(string);
```

Comment:
Why doing this `stringify` and `parse`?

Reply:
I have no idea honestly, it's there for reasons, but I can't remember anymore why

Reply:
Yeah not sure that this is 100 needed, we should be mutating these either. The only think I could think is that the `output` wasn't serializeable, but it seems to be fine?